network_diffusion/models/ICM_model.py

```diff
@@ -0,0 +1,173 @@
+            net=net,
+            actorwise=True
+        )
+        seed_nodes: List[NUBff] = []
+        for idx, actor in enumerate(self._seed_selector.actorwise(net=net)):
+            if idx < budget[self.PROCESS_NAME][self.ACTIVE_NODE]:
+                a_state = self.ACTIVE_NODE
+            else:
+                a_state = self.INACTIVE_NODE
+
+            for l_name in actor.layers:
+                seed_nodes.append(
+                    NUBff(
+                        node_name=actor.actor_id,
+                        layer_name=l_name,
+                        new_state=a_state
+                    )
+                )
+        out_json = self.update_network(net=net, activated_nodes=seed_nodes)
+        return out_json
+
+    def agent_evaluation_step(
+        self,
+        agent: MLNetworkActor,
+        layer_name: str,
+        net: MultilayerNetwork
+
+    ) -> str:
+        """
+
+        param_agent is node to check
+        param_layer_name is node's layer
+        param_net is choosing network
+
+        """
+
+        l_graph: nx.Graph = net.layers[layer_name]
+        current_state = l_graph.nodes[agent.actor_id]["status"]
+        if current_state == self.ACTIVE_NODE:
+            return self.ACTIVATED_NODE
+        # checking neighborhood of actor
+        for neighbour in l_graph.neighbors(agent.actor_id):
+            # if neighbour is active, it can send signal to activation of actor
+            if l_graph.nodes[neighbour]["status"] == self.ACTIVE_NODE:
+                # random is homogeneous distribution
+                r = random.random()
+                # actor is activated if a probability allow its
+                if r < self.probability:
+                    return self.ACTIVE_NODE
+        return current_state
+
+    def network_evaluation_step(
+        self, net: MultilayerNetwork
+    ) -> List[NUBff]:
+
+        """
+        param_net: load network
+        return a list of activated_nodes if the protocol allow
+
+        """
+
+        activated_nodes: List[NUBff] = []
+
+        for actor in net.get_actors():  # checking each node
+            layer_inputs = {}
+
+            if set(actor.states.values()) == set(self.ACTIVATED_NODE):
+                continue
+            for layer_name in actor.layers:
+                # downloading the status of actor
+                layer_inputs[layer_name] = self.agent_evaluation_step(actor, layer_name, net)
+            if self.protocol(layer_inputs):  # checking the protocol's conditions
+                # adding to list
+                activated_nodes.extend(
+                    [
+                        NUBff(
+                            node_name=actor.actor_id,
+                            layer_name=layer_name,
+                            new_state=self.ACTIVE_NODE,
+                        )
+                        for layer_name in layer_inputs
+                    ]
+                )
+        return activated_nodes
```

Comment:
you're applying protocol to actors that changed their state from active to activated, which is incorrect!